File: tron_unity/Assets/Scripts/Tron.cs
using Unity.VisualScripting;
using UnityEngine;

public class TronBike
{
    public Vector2Int pos;
    public TronBike(Vector2Int startPos)
    {
        pos = startPos;
    }

    // Returns true if crash
    public bool Move(Vector2Int vel, int[,] trails)
    {
        pos += vel;
        pos.x = Mathf.Clamp(pos.x, 0, trails.GetLength(0) - 1);
        pos.y = Mathf.Clamp(pos.y, 0, trails.GetLength(1) - 1);
        return IsHit(trails, pos.x, pos.y);
    }

    public bool IsHitInDir(int[,] trails, Vector2Int dir)
    {
        Vector2Int newPos = pos + dir;
        return IsHit(trails, newPos.x, newPos.y);
    }

    public bool IsHit(int[,] trails, int x, int y)
    {
        return trails[x, y] != 0;
    }
}

public enum GameState { Draw, Bike1Win, Bike2Win, Playing }
public class Tron
{
    public int width;
    public int height;

    public int[,] trails;

    public TronBike bike1;
    public TronBike bike2;

    public Tron(Vector2Int size)
    {
        width = size.x;
        height = size.y;
    }

    public void Reset()
    {
        trails = new int[width, height];
        bike1 = new TronBike(new Vector2Int(width / 6, height / 2));
        bike2 = new TronBike(new Vector2Int(5 * width / 6, height / 2));
    }

    // Returns:
    // -1 - No collision
    // 2 - Bike1 collided
    // 1 - Bike2 collided
    // 0 - Both bikes collided (draw)
    public GameState Step(Vector2Int dir1, Vector2Int dir2)
    {
        // Mark current positions as trails
        trails[bike1.pos.x, bike1.pos.y] = 1;
        trails[bike2.pos.x, bike2.pos.y] = 2;

        // Move bikes
        bool bike1Hit = bike1.Move(dir1, trails);
        bool bike2Hit = bike2.Move(dir2, trails);

        // Head-to-head collision
        if ((bike1Hit && bike2Hit) || bike1.pos == bike2.pos)
            return GameState.Draw;

        if (bike1Hit)
            return GameState.Bike2Win;

        if (bike2Hit)
            return GameState.Bike1Win;


        return GameState.Playing;
    }
}

File: rl_core/environment/tron.py
import numpy as np
from rl_core.utils.helper import bcolors

class Bike:
    
    def __init__(self, pos):
        self.pos = np.array(pos, dtype=np.int8)  # (x, y) !!!!

    def move(self, vel, walls):
        """Move bike if success and return True if crash"""
        new_pos = self.pos + vel
        if self.is_hit(walls, *new_pos):
            return True
        self.pos = new_pos

    def is_hit(self, walls, x, y):
        return not 0 <= y < len(walls) or not 0 <= x < len(walls[0]) or walls[y, x] != 0

class Tron:
    
    def __init__(self, size):
        self.width, self.height = size

    def reset(self):
        self.walls = np.zeros((self.height, self.width), dtype=np.int8)
        self.bike1 = Bike([1, self.height // 2])
        self.bike2 = Bike([self.width - 2, self.height // 2])
        self.walls[self.bike1.pos[1], self.bike1.pos[0]] = 1
        self.walls[self.bike2.pos[1], self.bike2.pos[0]] = 2

    def tick(self, dir1, dir2):
        """
        Updates the game state by moving the bikes and checking for collisions.
        
        :param dir1: Direction for bike1 as a tuple (dx, dy)
        :param dir2: Direction for bike2 as a tuple (dx, dy)
        :return: An integer indicating the result of the tick:
                 -1 - No collision
                 2 - Bike1 collided
                 1 - Bike2 collided
                 0 - Both bikes collided (draw)
        """
        assert type(dir1) == np.ndarray and type(dir2) == np.ndarray, f"{bcolors.FAIL}Invalid direction type : {type(dir1)}, {type(dir2)}{bcolors.ENDC}"
        assert dir1.shape == (2, ) and dir2.shape == (2, ), f"{bcolors.FAIL}Invalid direction shape : {dir1.shape}, {dir2.shape}{bcolors.ENDC}"
        
        # Move bikes
        bike1_hit = self.bike1.move(dir1, self.walls)
        bike2_hit = self.bike2.move(dir2, self.walls)

        if (bike1_hit and bike2_hit) or all(self.bike1.pos == self.bike2.pos):
            return 0  
        if bike1_hit:
            return 2  
        if bike2_hit:
            return 1  

        self.walls[self.bike1.pos[1], self.bike1.pos[0]] = 1
        self.walls[self.bike2.pos[1], self.bike2.pos[0]] = 2
        return -1
        
